# Problem Set Examples

This is a *Jupyter* notebook using interactive *Python*. 

We can use this notebook as a calculator. The benefit is that all **our calculations can be documented** so that we can easily find errors or make changes without starting from the beginning. I recommend using *Python* notebooks for all your math. There are thousands of tools available through *Python* and we can go from the simple math of this notebook all the way to powerful data science and AI projects.

We will demonstrate the problems from the "**Math Ideas for Biochemistry - Part 1**" document available on the moodle site. Observe how the *Python* code serves as the information on the method used. Now you can check my work and fix my mistakes. I make no promises that these answers are correct. It is your job to confirm them.

## Problem 1 from Math Ideas

The molecular mass of glucose is $180.2 \frac{g}{mole}$. A sugar cube has a mass of $4.3$ grams. What is the concentration of a solution of glucose when 3 sugar cubes are dissolve in a total volume of $150$ mL of water? 

**Solution**

This is a straightforward concentration problem. Observe the answer in the code below.

In [49]:
mw = 180.2   #### g/mole
m = 4.3      #### g
v = 0.150    #### L

moles = m / mw
conc = moles / v

print(f"The concentration of glucose is {conc:0.2f} moles/L")

### THIS ANSWER IS WRONG. READ THE QUESTION CAREFULLY AND FIND MY ERROR. 
### The fact that my math was documented as python code allowed you to
### find my mistake before the spaceship crashed. Small errors are just as bad 
### as big errors when you want to go to Mars. Both will result in your
### billion-dollar spaceship being in a billion pieces.
### There are no part marks on Mars.

The concentration of glucose is 0.16 moles/L


## Problem 2 from Math Ideas

The equilibrium constant, $\rm K_{eq}$, for $\beta$-glucose converting to $\alpha$-glucose is $0.56$. I observe that the concentration of $\alpha$-glucose in a sample is $1.2$ mM. What is the concentration of $\beta$-glucose if this system has reached equilibrium?

**Solution**

Use the equilibrium constant to solve the problem. The answer is documented in the code below. I don't have to explain anything as every step is described in the code. That is one great strength of recording you math in a language like *Python*.


In [51]:
Keq = 0.56
conc_alpha_glucose = 0.0012 # moles/L

### Equilibrium equation rearranged to solve for beta glucose

conc_beta_glucose = conc_alpha_glucose / Keq

print(f"The concentration of beta-glucose at equilibrium will be {conc_beta_glucose:0.2e} moles/L")
print(f"The concentration of beta-glucose at equilibrium will be {conc_beta_glucose:0.9f} moles/L")
print(f"The concentration of beta-glucose at equilibrium will be {conc_beta_glucose:0.3g} moles/L")
print(f"The concentration of beta-glucose at equilibrium will be {1000 * conc_beta_glucose:0.2f} millimoles/L")

### What have you learned about printing numbers from the code above? Delete the "f" at the start of 
### the print command and see what happens. Change the numbers in :0.2f, etc. What does :10.10f do?



The concentration of beta-glucose at equilibrium will be 2.14e-03 moles/L
The concentration of beta-glucose at equilibrium will be 0.002142857 moles/L
The concentration of beta-glucose at equilibrium will be 0.00214 moles/L
The concentration of beta-glucose at equilibrium will be 2.14 millimoles/L


## Problem 3 from Math Ideas

What is the free energy potential for a mixture of $\beta$-glucose and $\alpha$-glucose at equilibrium?

**Solution**

By definition, when we are at equilibrium the difference in energy between the population of products and the population of reactants is zero.

## Problem 4 from Math Ideas

What is the free energy potential, $\Delta G$, of a solution of $0.1$ M $\alpha$-glucose and $0.4$ M $\beta$-glucose? If we are not at equilibrium, which direction will the process move to reach equilibrium. What would be the concentrations in this system after it reaches equilibrium? If I add a catalyst, how will it change the equilibrium?

**Solution**

The free energy of a system, $\Delta G$, depends on the difference in free energy at the standard state, $\Delta G^{\circ}$, and the concentrations of reactants (the reaction quotient, $Q$)

$$\Delta G = \Delta G^{\circ} + RT\ln Q$$ 

where 

$$Q = \frac{[products]}{[reactants]}$$

First obtain the standard free energy difference, $\Delta G^{\circ}$, using the equilibrium constnat reported above. The use the concentrations observed in this case to calculate the free energy difference, $\Delta G$, in these conditions. When temperature is not reported always use $25$ $^{\circ} C$.

If the answer below is correct (check my work!) then the system is not at equilibrium and will proceed toward product as it moves to reach equilibrium.

In [29]:
import numpy as np

### get standard free energy difference from Keq in problem 2

Keq = 0.56   ### note that Q = Keq when we are at equilibrium
d_G = 0      ### the free energy difference is zero at equilibrium

R = 8.314  # J / (K * mole) gas constant
T = 273 + 25  # 25 degrees C in kelvin = 298K

### rearrange the equation to solve for standard free energy. 

Q = Keq
d_G_standard = -R*T*np.log(Q)

print(f"The standard difference in free energy is {d_G_standard/1000:+0.2f} kJ/mole")

conc_alpha_glucose = 0.1  ### moles/L
conc_beta_glucose = 0.4   ### moles/L

Q = conc_alpha_glucose / conc_beta_glucose  ### alpha is product, beta is reactant

d_G = d_G_standard + R * T * np.log(Q)

print(f"The difference in observed free energy at the concentrations given is {d_G/1000:+0.2f} kJ/mole")

The standard difference in free energy is +1.44 kJ/mole
The difference in observed free energy at the concentrations given is -2.00 kJ/mole


## Problem 5 from Math Ideas

The specific optical rotation of $+\alpha$-glucose is $112\frac{degrees}{dm\cdot M}$. What is the observed optical rotation observed using a polarimeter for a solution with a concentration of $5.3$ mM? Assume that we are using the standard $10$ cm (one dm) sample tube.

**Solution**

This is an example of straightforward dimensional analysis. Examine the code below and correct any errors.

In [30]:
sp_opt_rotation = 112   ### degrees per decimeter per molar
conc = 5.3              ### millimolar. We will need to convert this to molar
conc = conc / 1000      ### we convert millimoles to moles to get mole/L
d = 1                   ### decimeters. often ignored in these calculations

obs_rotation = sp_opt_rotation * conc * d

print(f"The observed optical rotation is {obs_rotation:+0.2f} degrees")

The observed optical rotation is +0.59 degrees


## Problem 6 from Math Ideas

How many microlitres are there in a millilitre? How many grams are there in a kilogram? What is $10$ $\mu M$ in units of Molar? What is $7$ $\frac{kJ}{mole}$ in units of $\frac{J}{mole}$?

**Solution**

This is all unit conversions. I will leave you to answer this on your own.


## Problem 7 from Math Ideas

I am observing the initial rate of a reaction at a wavelength of $280$ nm. The molar extinction coefficient, $e_{280}$, of the starting material at $280$ nm is $7000$ $M^{-1}$ (for a $1$ cm cuvette). The initial absorbance of the reaction mixture is $0.87$ AU. After one minute, the absorbance is $0.82$ AU. (Assume that our reactant is the only molecule that absorbs light at $280$ nm).

1. What is the initial concentration of the starting material?
1. What is the concentration after one minute?
1. What is the initial rate of reaction based on those two data points?
1. What is the first order rate constant for the disappearance of the observed reactant (assuming a first order reaction)?
1. Is the initial rate method appropriate in this case?

**Solution**

Examine the code below for the answers. I am presenting each part of the question is a separate block of code. 

In [41]:
e_280 = 7000  ### per molar
A_0 = 0.87    ### initial absorbance

conc_0 = A_0 / e_280

print(f"The initial concentration is {conc_0:0.9f} moles/L")
print(f"The initial concentration is {conc_0 * 1000:0.4f} millimoles/L")

The initial concentration is 0.000124286 moles/L
The initial concentration is 0.1243 millimoles/L


In [40]:
A_t = 0.82     ### absorbance at time point (one minute)

conc_t = A_t / e_280

print(f"The concentration after one minute is {conc_t * 1000:0.4f} millimoles/L")

The concentration after one minute is 0.1171 millimoles/L


In [43]:
time = 1       ### time is 1 minute

init_rate = (A_t - A_0) / time

print(f"The initial rate of change in concentration is {init_rate * 1000:0.2f} millimoles/L.min")

The initial rate of change in concentration is -50.00 millimoles/L.min


In [48]:
first_order_rate_const = -init_rate / A_0  ### initial rate / initial concentration

print(f"The rate constant is {first_order_rate_const} per minute")
print(f"The rate constant is {first_order_rate_const:0.3f} per minute")
print(f"The rate constant is {first_order_rate_const:0.3g} per minute")
print(f"The rate constant is {first_order_rate_const:0.3e} per minute")
print(f"The rate constant is {first_order_rate_const/60:0.3e} per second")

The rate constant is 0.05747126436781615 per minute
The rate constant is 0.057 per minute
The rate constant is 0.0575 per minute
The rate constant is 5.747e-02 per minute
The rate constant is 9.579e-04 per second


## Problem 8 from Math Ideas

I add 10 mL of 0.1 M HCl to 170 mL of pure water. What is the *p*H of this new solution? 

**Solution**

This is a simple $\rm c_1v_1 = c_2v_2$ problem with the addition of converting your answer to $\rm pH$.

The initial colume, $\rm c_1$, is 10 mL. The final volume, $\rm C_2$, will be 180 mL (170 mL + 10 mL). The initial concentration of HCl is 0.1 M. HCl is a very strong acid and will be 100% dissociated, so we can say that the hydronium ion concentration will be the concentration of HCl.

Examine the *Python* commands below. Observe how they are no different than using a calculator.

In [1]:
import numpy as np    # import additional math tools

v1 = 10 # mL
v1 = v1 / 1000  # convert mL to L
c1 = 0.1 # moles/L

volumeWater = 170 # mL

v2 = volumeWater + v1
v2 = v2 / 1000  # convert mL to L

c2 = v1 * c1 / v2

pH = -np.log10(c2)

print(f"the pH is {pH:0.2f}")

the pH is 2.23


## Problem 9 from Math Ideas

I have 230 mL of a solution of pure water with enough HCl added to make the *p*H equal to 4.5. I add 10 mL of a 0.15 M NaOH solution. What is the *p*H now? 

**Solution**

We have a volume and concentration for an HCl solution. $\rm [HCl] = [H^+]$ and we can get $\rm [H^+]$ from the $\rm pH$ value.

Obtain the number of moles of HCl present, subtract the number of moles of NaOH added, and then divide by the new volume to get the concentration of HCl remaining. Then convert to $\rm pH$.

In [2]:
pH = 4.5
v1 = 230 # mL
v1 = v1/1000 # convert mL to L
c1 = 10**(-pH) # Units will be moles/L

moles_HCl = c1 * v1

volumeNaOHsolution = 10 # mL
volumeNaOHsolution = volumeNaOHsolution/1000 # convert mL to L
concNaOHsolution = 0.15 # moles/L

molesNaOH = volumeNaOHsolution * concNaOHsolution

molesBase = molesNaOH - moles_HCl

concBase = molesBase / (v1 + volumeNaOHsolution)

pOH = -np.log10(concBase)
pH = 14 - pOH

print(f"the pH is {pH:0.2f}")

the pH is 11.79


## Problem 10 from Math Ideas

I have 230 mL of a solution of 0.2 M acetic acid and have added enough sodium hydroxide solution to set the $\rm pH$ to 4.5. The $\rm pK_a$ of acetic acid is 4.7. 

1. What are the concentrations of acetate ion and acetic acid in this solution? 
2. I add 10 mL of a 0.15 M NaOH solution. What is the $\rm pH$ now? 

**Solution**

Calculate the moles of Acetic acid (the acidic for, HOAc) in the sample and compare it to the moles of NaOH to be added.  We will see that there will be an excess of NaOH. So I subtract [HOAc] from [NaOH] to get the concentration of hydroxide ion in the new solution. I then obtained the $\rm pOH$, and from that, the $\rm pH$


In [3]:
pHsolution = 4.5
volumeSolution = 230 # mL
volumeSolution = volumeSolution/1000 # convert mL to L
concAceticAcid = 0.2 # moles/L
molesAceticAcid = concAceticAcid * volumeSolution

pKaAceticAcid = 4.7

# Use Henderson-Hasselbalch
# pH = pKa + log10(A/HA)
#
# A/HA = 10**(pH - pKa)

A_HA_ratio = 10**(pHsolution - pKaAceticAcid)
print(f"the A/AH ratio is {A_HA_ratio:0.3f}")

moleFractionA = A_HA_ratio/(A_HA_ratio+1)
moleFractionHA = 1/(A_HA_ratio+1)

volumeNaOHsolution = 10 # mL
volumeNaOHsolution = volumeNaOHsolution/1000 # convert mL to L
concNaOHsolution = 0.15 # moles/L

molesNaOH = volumeNaOHsolution * concNaOHsolution
print(f"molesNaOH = {molesNaOH:0.4f}")


molesA = molesAceticAcid * moleFractionA
molesHA = molesAceticAcid * moleFractionHA

print(f"molesA = {molesA:0.3f}")
print(f"molesHA = {molesHA:0.3f}")
print(f"It will take {molesHA/concNaOHsolution*1000:0.1f} mL of NaOH solution to overcome the buffer completely")

molesA = molesA + molesNaOH
molesHA = molesHA - molesNaOH

print(f"After NaOH molesA = {molesA:0.3f}")
print(f"After NaOH molesHA = {molesHA:0.3f}")

A_HA_ratio = molesA/molesHA
print(f"After NaOH the A/AH ratio is = {A_HA_ratio:0.3f}")


pH = pKaAceticAcid + np.log10(A_HA_ratio)
print()
print(f"the pH is {pH:0.2f}")

the A/AH ratio is 0.631
molesNaOH = 0.0015
molesA = 0.018
molesHA = 0.028
It will take 188.0 mL of NaOH solution to overcome the buffer completely
After NaOH molesA = 0.019
After NaOH molesHA = 0.027
After NaOH the A/AH ratio is = 0.723

the pH is 4.56


## Making a Buffer for the Assignment

The math below involves making a buffer for a biochemical experiment. In this example, I will be using HEPES buffer. It is an organic molecule that is compatible with cells and the acidic form has a $\rm pK_a$ value of $7.43$. I will purchase a bottle of crystaline [HEPES (the acidic form)](https://www.sigmaaldrich.com/CA/en/substance/hepes238307365459) and a bottle of crystaline [HEPES sodium salt (the basic form)](https://www.sigmaaldrich.com/CA/en/substance/hepessodiumsalt2602975277393). I want to make a $1.5$ L volume with a total HEPES concentration of $200$ mM and a $\rm pH$ of $7.15$.

We know the $\rm pH$, the $\rm pK_a$ of the acid form of HEPES and the molar mass of both materials (just click on the links in the above paragraph to get more information about both).

We can get the needed ratio of the two forms, the value of $\frac{[A]}{[HA]}$ required if $\rm pH = 7.15$ by using the Hendersson-Haselbalch equation. 

$$pH = pK_a + \log{\frac{[A]}{[HA]}}$$

and we want the total concentration of both forms of HEPES to be $200$ mM so we can say that

$$[A]+[HA]=0.200$$

The code below will serve as a calculator for our math.  Observe how what you do is recorded in the text of the code itself. You can come back to this code and change some numbers to use it again. Feel free to **steal it** and modify it for your own needs.



In [14]:
#######################################################################
#### Set up variables. Change these if I want something different
#######################################################################

pH = 7.15             #### chosen pH
pKa = 7.45            #### pKa value of acid form of buffer. I looked it up on wikipedia
  
MolMass_HA = 238.30   #### g/mole. Molar mass of acidic form. I looked it up in the specification sheet available from the vendor
MolMass_A = 260.29    #### g/mole. Molar mass of sodium salt of HEPES (basic form). I looked it up in the specification sheet available from the vendor
  
Total_Conc = 0.200    #### mole/L.  Desired total concentration.
  
Volume = 1.5          #### L. Volume of buffer solution.

#######################################################################
#### Calculate A/HA ratio
#######################################################################

ratio_A_HA = 10 ** (pH - pKa)            ### Reorganized HH eqn to solve for A/HA

print(f"Ratio of A/HA is {ratio_A_HA:0.3f}")
print()

#######################################################################
#### Calculate A and HA concentrations
#######################################################################

r = ratio_A_HA       #### so that the code below will be easier to read

percent_A =  r / (1 + r)  #### Convert ratio to mole fraction (percent)
percent_HA = 1 / (1 + r)

conc_HA = Total_Conc * percent_HA   #### the fraction of total conc is the conc of each.
conc_A = Total_Conc * percent_A

print(f"Concentration of A is {conc_A:0.3f} moles/L")
print(f"Concentration of HA is {conc_HA:0.3f} moles/L")

print(f"Total concentration of HA + A is {conc_HA+conc_A:0.3f} moles/L")   #### Check. Just in case.
print()

#######################################################################
#### Calculate A and HA amounts
#######################################################################

mass_of_A = conc_A * Volume * MolMass_A       #### moles = conc * volume.  mass = moles * molar_mass
mass_of_HA = conc_HA * Volume * MolMass_HA

print(f"Mass of A in {Volume:0.1f}L is {mass_of_A:0.2f} grams")
print(f"Mass of HA in {Volume:0.1f}L is {mass_of_HA:0.2f} grams")



Ratio of A/HA is 0.501

Concentration of A is 0.067 moles/L
Concentration of HA is 0.133 moles/L
Total concentration of HA + A is 0.200 moles/L

Mass of A in 1.5L is 26.07 grams
Mass of HA in 1.5L is 47.62 grams


### Making the Buffer

We see from the math above that I will need $47.62$ grams of the HEPES acid and $26.07$ grams of its sodium salt dissolved in a final volume of $1.5$ litres. Now I just have to write up a recipe. Take note of how the code above is just calculation steps that are recorded as *Python* commands and can be checked and corrected easily by the reader. Try seeing what the calculations are in a printout of an Excel preadsheet. How does this compare.

In you recipe you will describe your calculations using the math typesetting tool of your choice, such as the equation editor built into your word processor. *Python* is your calculator, not your word processor.